pkg/azure/utils.go

```diff
@@ -67,20 +75,133 @@ func SetupAzureTesting(subscriptionId string, testName string) string {
 	return resourceGroupName
 }
 
-func TeardownAzureTesting(subscriptionId string, resourceGroupName string) {
+func TeardownAzureTesting(subscriptionId string, resourceGroupName string, namespace string) {
 	if os.Getenv("INVISINETS_TEST_PERSIST") != "1" {
-		ctx := context.Background()
-		resourceGroupsClient := createResourceGroupsClient(subscriptionId)
-		poller, err := resourceGroupsClient.BeginDelete(ctx, resourceGroupName, nil)
+		if os.Getenv("INVISINETS_AZURE_RESOURCE_GROUP") == "" {
+			// Delete resource group
+			ctx := context.Background()
+			resourceGroupsClient := createResourceGroupsClient(subscriptionId)
+			poller, err := resourceGroupsClient.BeginDelete(ctx, resourceGroupName, nil)
+			if err != nil {
+				// If deletion fails: refer to https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/delete-resource-group
+				panic(fmt.Sprintf("Error while deleting resource group: %v", err))
+			}
+			_, err = poller.PollUntilDone(ctx, nil)
+			if err != nil {
+				panic(fmt.Sprintf("Error while waiting for resource group deletion: %v", err))
+			}
+		} else {
+			// Delete resources without deleting the resource group
+			// Order deletion is based off of https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/delete-resource-group#how-order-of-deletion-is-determined
+			ctx := context.Background()
+			cred, err := azidentity.NewDefaultAzureCredential(nil)
+			if err != nil {
+				panic(fmt.Sprintf("Error while getting azure credentials: %v", err))
+			}
+			resourcesClient, err := armresources.NewClient(subscriptionId, cred, nil)
+			if err != nil {
+				panic(fmt.Sprintf("Error while creating resources client: %v", err))
+			}
+			providersClient, err := armresources.NewProvidersClient(subscriptionId, cred, nil)
+			if err != nil {
+				panic(fmt.Sprintf("Error while creating providers client: %v", err))
+			}
+			pager := resourcesClient.NewListPager(&armresources.ClientListOptions{
+				Filter: proto.String(fmt.Sprintf("tagName eq '%s' and tagValue eq '%s'", namespaceTagKey, namespace)),
+			})
+			resourceTypeToResources := make(map[string][]*armresources.GenericResourceExpanded)
+			resourceTypeToAPIVersion := make(map[string]string)
+			for pager.More() {
+				result, err := pager.NextPage(ctx)
+				if err != nil {
+					panic(fmt.Sprintf("failed to get next page of resources: %v", err))
+				}
+
+				for _, resource := range result.Value {
+					resourceIDInfo, err := getResourceIDInfo(*resource.ID)
+					if err != nil {
+						panic(fmt.Sprintf("Unable to parse resource ID info: %v", err))
+					}
+
+					// Need to check here since resourceGroup can't be used with tags in the above filter
+					// Ignore case when checking because Azure sometimes capitalizes the resource group name for no apparent reason
+
+					if strings.EqualFold(resourceIDInfo.ResourceGroupName, resourceGroupName) {
+						if _, ok := resourceTypeToResources[*resource.Type]; !ok {
```

Comment:
Nit: Would have helped me reading this to have a comment like "Group resources by type and get the correct API version for the type" or something like that